packages/next/src/build/index.ts

```diff
@@ -349,6 +350,34 @@ function pageToRoute(page: string) {
   }
 }
 
+function createAppDataRouteInfo(
```

Comment:
This function seems odd to me

experimentalPPR will always be false when isRouteHandler is true. So having two boolean flags seems suspect already. My guess is you should have two functions one for route handlers and one for renders. Then just call the right one based on the situation. You can also then just read from config to determine if you need the PPR prefetch route

Reply:
Unfortunately we use the same larger `build` function here to build each of the pages (pages and routes included), so the point of extracting this was to allow the "app data routes" to be defined by the dependencies (`isRouteHandler` and `experimentalPPR`).

> You can also then just read from config to determine if you need the PPR prefetch route

`experimentalPPR` means something specific here, it means that this particular route is enabled for PPR. I'm currently looking into cases like when there's a segment config of `dynamic = "force-static"` that should actually ensure that the route doesn't opt into PPR behaviors, and should instead be outputted on infrastructure as a purely static route.

So technically having this function here extracting this logic is preparing us for that particular distinction, which I think is important for documenting.

Reply:
Does a route handler having a dataRoute or prefetchDataRoute ever make sense?

Maybe my issue is with the polymorphism of the route type. Reading the code it is really hard to track that a route handler will always have a null/undefined value for these two items.

Something like this reads more verbosely but IMO much clearer in intention and behavior.

```
let dataRoute, prefetchDataRoute;
if (isRouteHandler) {
  // Route Handlers never have a dataRoute or pefetchDataRoute
  dataRoute = null;
  prefetchDataRoute = undefined;
} else {
  // render pages always have a dataRoute and if PPR is enabled for this page or route segment
  // it must also have prefetchDataRoute
  const normalizedRoute = normalizePagePath(...)
  dataRoute = dataRouteFromNormalizedRoute(normalizedRoute);
  
  if (experimentalPPR) {
    prefetchDataRoute = prefetchDataRouteFromNormalizedRoute(normalizedRoute);
  } else {
    prefetchDataRoute = undefined;
  }
}
```

The fact that this needs to be duplicated in a few places is unfortunate but I think that could be solved by forking the route handler treatment and render handlers at a higher level. By putting that inside `createAppDataRouteInfo` you obuscate the fact that routeHandlers will never have these routes